tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN, destination);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_006 [ interlocked_compare_exchange shall compare the 32-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.*]
+Tests_SRS_INTERLOCKED_43_037: [ interlocked_compare_exchange shall return the initial value of *destination.]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_not_equal_case)
+{
+    ///arrange
+    volatile int32_t destination = INT32_MAX;
+    int32_t comperand = INT32_MIN;
+    int32_t exchange = INT32_MIN;
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, destination);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_007 [ interlocked_compare_exchange_128 shall compare *destination with *comperand_result. If they are equal, destination[0] is set to exchange_low and destination[1] is set to exchange_high. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_039: [ interlocked_compare_exchange_128 shall store the initial value of *destination in *comperand_result regardless of the result of the comparison.` ]
+Tests_SRS_INTERLOCKED_43_038: [ interlocked_compare_exchange_128 shall return true if *comperand_result equals the original value of *destination.]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_128_equal_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+    int64_t* comperand_result = (int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+
+    destination[0] = INT64_MAX;
+    destination[1] = INT64_MIN;
+    comperand_result[0] = INT64_MAX;
+    comperand_result[1] = INT64_MIN;
+
+    int64_t exchange_high = 2;
+    int64_t exchange_low = 3;
+
+    ///act
+    bool return_val = interlocked_compare_exchange_128(destination, exchange_high, exchange_low, comperand_result);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, 3, destination[0]);
+    ASSERT_ARE_EQUAL(int64_t, 2, destination[1]);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, comperand_result[0]);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, comperand_result[1]);
+    ASSERT_IS_TRUE(return_val);
+
+    ///cleanup
+    _aligned_free((void*)destination);
+    _aligned_free(comperand_result);
+}
+
+
+/*
+Tests_SRS_INTERLOCKED_43_039: [ interlocked_compare_exchange_128 shall store the initial value of *destination in *comperand_result regardless of the result of the comparison.` ]
+Tests_SRS_INTERLOCKED_43_063: [ interlocked_compare_exchange_128 shall return false if *comperand_result does not equal the original value of *destination. ]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_128_not_equal_case)
```

Comment:
>_case [](start = 56, length = 5)

Note on test names: Some time ago I read "The Art of unit testing" also known as the samurai book by Roy Osherove. Maybe you heard of it. If not it might be an interesting read. One of the things that stuck with me from that book were several rules/patterns for naming tests and how to write good tests.
One of them is that the test names should express the requirement. The reason to be verbose is to be able to immediately see in the test name what is there to be tested by that test function.

Thus I tend to name my tests like: interlocked_compare_exchange_128_does_not_exchange_when_not_equal

It's not that different from what you had, but as tests become more complex it is I think important to pay attention to what the name of the test is. Usually having the verb in it forces you to have a (slightly) better test name.